app/src/main/java/org/oppia/app/customview/interaction/NumberWithUnitsInputInteractionView.kt

```diff
@@ -0,0 +1,245 @@
+package org.oppia.app.customview.interaction
+
+import android.content.Context
+import android.util.AttributeSet
+import android.widget.EditText
+import org.json.JSONObject
+import org.oppia.app.model.Fraction
+import org.oppia.app.model.InteractionObject
+import org.oppia.app.model.NumberUnit
+import org.oppia.app.model.NumberWithUnits
+import org.oppia.app.model.UserAnswer
+import org.oppia.app.parser.StringToFractionParser
+import org.oppia.app.player.state.answerhandling.InteractionAnswerHandler
+import org.oppia.domain.util.normalizeWhitespace
+import java.lang.Double.parseDouble
+import java.util.*
+import java.util.regex.Pattern
+
+/** The custom EditText class for number input with units interaction view. */
+class NumberWithUnitsInputInteractionView @JvmOverloads constructor(
+  context: Context,
+  attrs: AttributeSet? = null,
+  defStyle: Int = android.R.attr.editTextStyle
+) : EditText(context, attrs, defStyle), InteractionAnswerHandler {
+  private lateinit var type: String
+  private var real: Float = 0f
+  private lateinit var fractionObject: Fraction
+  private lateinit var value: String
+  private lateinit var units: String
+  private lateinit var CURRENCY_UNITS: JSONObject
+  /** @return index of pattern in s or -1, if not found
+   */
+  fun indexOf(pattern: Pattern, s: String): Int {
+    val matcher = pattern.matcher(s)
+    return if (matcher.find()) matcher.start() else -1
+  }
+
+  override fun getPendingAnswer(): UserAnswer {
+    val userAnswerBuilder = UserAnswer.newBuilder()
+    if (text.isNullOrEmpty() || isValidNumberWithUnits()) {
+      return userAnswerBuilder.build()
+    }
+
+    val answerTextString = text.toString()
+    userAnswerBuilder.answer = InteractionObject.newBuilder()
+      .setNumberWithUnits(getNumberWithUnits(text.toString()))
+      .build()
+    userAnswerBuilder.plainAnswer = answerTextString
+
+    return userAnswerBuilder.build()
+  }
+
+  fun getCurrencyUnits(): JSONObject {
+    return JSONObject(
```

Comment:
I don't think it makes sense to keep this data in the form of a JSON object. Either we should move it to be an asset that's bundled with the APK (e.g. if we're planning to configure it), move it over to being a response from the server (if we want to remotely change it for already deployed apps), or hardcode it into the application. As-is, we don't need to do any JSON parsing here if we already have all the data in-memory.

For Oppia web, where does this object come from? Is it provided by the server? I think we either need to make a network request to retrieve this file (in which case we need to create a controller for it), or this should be hardcoded directly in the new parsing utility class.

Reply:
moved to assets